iroh-net/src/ping.rs

```diff
@@ -17,32 +33,59 @@ impl Debug for Inner {
     }
 }
 
+#[derive(Default)]
 struct Inner {
-    client_v6: Client,
-    client_v4: Client,
+    client_v6: Mutex<Option<Client>>,
+    client_v4: Mutex<Option<Client>>,
 }
 
 const DEFAULT_TIMEOUT: Duration = Duration::from_secs(5);
 
 impl Pinger {
     /// Create a new [Pinger].
-    pub fn new() -> Result<Self> {
-        let client_v4 = Client::new(&Config::builder().kind(ICMP::V4).build())
-            .context("failed creating IPv4 pinger")?;
-        let client_v6 = Client::new(&Config::builder().kind(ICMP::V6).build())
-            .context("failed creating IPv6 pinger")?;
-
-        Ok(Self(Arc::new(Inner {
-            client_v4,
-            client_v6,
-        })))
+    pub fn new() -> Self {
+        Self(Arc::new(Default::default()))
+    }
+
+    /// Lazily create the ping client.
+    ///
+    /// We do this because it means we do not bind a socket until we really try to send a
+    /// ping.  It makes it more transparent to use the pinger.
+    fn get_client(&self, kind: ICMP) -> Result<Client> {
+        let client = match kind {
+            ICMP::V4 => {
+                let mut opt_client = self.0.client_v4.lock().unwrap();
```

Comment:
is it a good idea to use a sync mutex inside the async call stack from netcheck?

Reply:
This is without an await while holding the lock so the way tokio recommends it AFAIK.

The only other places this lock is held is in the other probes running at the same time.  In practice this will be the same probe for a different derper.  I think they're fine with blocking on each other (they run as a tokio TaskSet currently, though used to be FuturesUnordered before we figured out that panicked - in either case I still think that's fine).

Reply:
Well the taskset doesn't stop the cpu core to be blocked, when a mutex is held, so I don't think this will work out (FuturesUnordered would have the same issues)

Reply:
I'm not sure I follow.  Two cpu cores get to this mutex, one acquires it and one blocks for a short time after which it acquires it as well.  I don't think this does any harm for the tiny amount of non-blocking code bing run while holding the lock.

Reply:
If both tasks are running on the same core nothing will ever block I think.  Both tasks would get here, find the lock unlocked, lock, create or get the client, and unlock and carry on.  It would only be an issue if there's a lock held across an await point.

Is there another easy and better way to lazily create these clients that I am missing?

Reply:
you are right, sorry 😅 this stuff is hard